crates/sui-bridge-indexer-alt/src/handlers/token_transfer_handler.rs

```diff
@@ -0,0 +1,282 @@
+                tx.events.iter().flat_map(|events| &events.data).try_fold(
+                    results,
+                    |mut results, ev| {
+                        if self.deposited_event_type == ev.type_ {
+                            info!("Observed Sui Deposit {:?}", ev);
+                            // todo: metrics.total_sui_token_deposited.inc();
+                            results.push(TokenTransferDataWrapper::from_deposit_event(
+                                bcs::from_bytes(&ev.contents)?,
+                                tx,
+                                block_height,
+                                timestamp_ms,
+                            ));
+                        } else if self.approved_event_type == ev.type_ {
+                            info!("Observed Sui Approval {:?}", ev);
+                            // todo: metrics.total_sui_token_transfer_approved.inc();
+                            results.push(TokenTransferDataWrapper::from_approve_event(
+                                bcs::from_bytes(&ev.contents)?,
+                                tx,
+                                block_height,
+                                timestamp_ms,
+                            ));
+                        } else if self.claimed_event_type == ev.type_ {
+                            info!("Observed Sui Claim {:?}", ev);
+                            // todo: metrics.total_sui_token_transfer_claimed.inc();
+                            results.push(TokenTransferDataWrapper::from_claimed_event(
+                                bcs::from_bytes(&ev.contents)?,
+                                tx,
+                                block_height,
+                                timestamp_ms,
+                            ));
+                        }
+
+                        Ok(results)
+                    },
+                )
+            })
+    }
+}
+
+#[async_trait]
+impl Handler for TokenTransferHandler {
+    type Store = Db;
+
+    async fn commit<'a>(
+        values: &[Self::Value],
+        conn: &mut <Self::Store as Store>::Connection<'a>,
+    ) -> sui_indexer_alt_framework::Result<usize> {
+        let (token_transfers, data) = values.iter().fold(
+            (vec![], vec![]),
+            |(mut token_transfers, mut data_vec),
+             TokenTransferDataWrapper {
+                 token_transfer,
+                 data,
+             }| {
+                token_transfers.push(token_transfer.clone());
+                if let Some(d) = data {
+                    data_vec.push(d.clone());
+                }
+                (token_transfers, data_vec)
+            },
+        );
+
+        conn.transaction(|conn| {
+            async move {
+                {
+                    use token_transfer::columns::*;
+                    diesel::insert_into(token_transfer::table)
+                        .values(&token_transfers)
+                        .on_conflict((chain_id, nonce, status))
```

Comment:
This pipeline is concurrent, meaning it could commit checkpoint results out of order -- is that okay? It should be as long as you only expect one record here per key `(chain_id, none, status)`, and the conflict handling is only intended to deal with multiple writers writing the same records. If you need the conflict logic to deal with actually updating the same record in the table in response to multiple on-chain events, then it's going to be a problem.

Reply:
I think we can switch this to sequential pipeline, the old indexer framework is sequential and I forgot to handle that here.

Reply:
Actually `(chain_id, none, status)` is unique for Sui, the on conflict update is only for ETH data where a transaction can be not final....